### ISSUE TITLE
off by one in `#[n:m]pre`

### ISSUE THREAD

**Issue by chdoc** [EasyCrypt/easycrypt#109]
The following code shows that there is something wrong with the computation of "slices" from the precondition: 
```
require import AllCore.

module M = { 
  proc p (i : int) : int = {
    var j <- i;
    
    j <- j+1;
    return j; 
  }
}.

equiv foo :  M.p ~ M.p : ={arg} ==> ={res}.
conseq (: ={i} /\ true /\ 4 <= 4 ==> ={res}).
proc. 
(* seq 1 1 : #[1]pre.   selects ={1} *)
(* seq 1 1 : #[2]pre.   selects true *)
(* seq 1 1 : #[3]pre.   selects 4 <= 4*)
(* seq 1 1 : #[1:2]pre. selects ={1} only  *)
(* seq 1 1 : #[1:3]pre. selects ={1} /\ true only *)
```

**Comment by strub** [comment]
IMO, in slices, the last index is excluded.

**Comment by chdoc** [comment]
Well, IIUC, this means that there is no way to select the last conjunct in a slice, because in the above example `4` is an invalid index. I assume this is not desirable behavior.

**Comment by chdoc** [comment]
Also, since we start numbering conjuncts from 1 (rather than 0),  I think it would more natural to not do a "-1" on the second index, even if `4` where to become a valid index in the example above.

**Comment by fdupress** [comment]
@chdoc Actually, you can grab the last element in the slice using `#[n:]pre`. This mimics the slice syntax in python except for two facts:
- Python slices are 0-indexed, and
- Python slices don't seem to check bounds.

@strub Is the current behaviour expected? (Where the only way of defining a slice that contains the last conjunct is to use a right-open slice.)

**Comment by strub** [comment]
No, the behaviour seems rather clunky.

**Comment by fdupress** [comment]
This was fixed when merging #113 . Closing. (I should have used the commit message, sorry.)
